[Bug]: When applying an aws_cloudformation_stack_set_instance, i got an error: panic: unexpected format for ID parts ([<stackset_name>  <stack_set_instance>]), the following id parts indexes are blank ([1])

### Terraform Core Version

1.5.7

### AWS Provider Version

~> 5.0, 5.62.0, 5.49.0

### Affected Resource(s)

aws_cloudformation_stack_set_instance

### Expected Behavior

Terraform must deploy the stack_set_instance to account's list, without errors.

### Actual Behavior

Actually, when terraform just create the stack_set in the root account, but when will perform the application of the stack_set_instances, in each account informed on the list of accounts on the resource **aws_cloudformation_stack_set_instance**, an error appear and the apply simple fail!

### Relevant Error/Panic Output Snippet

```shell
running "## check diferent paths to terraform in atlantis\nif [ -f \"/atlantis-data/bin/terraform$ATLANTIS_TERRAFORM_VERSION\" ]\nthen\n  /atlantis-data/bin/terraform$ATLANTIS_TERRAFORM_VERSION apply  $PLANFILE \nelse\n  /usr/local/bin/terraform$ATLANTIS_TERRAFORM_VERSION apply  $PLANFILE\nfi\n\nif [ $? -eq 0 ]; then\n  export GITHUB_DEPLOYMENT_STATUS=success\n  /opt/deployments\n  sh /opt/ddog-logs.sh >/dev/null 2>&1\nelse\n  export GITHUB_DEPLOYMENT_STATUS=failure\n  /opt/deployments\n  sh /opt/ddog-logs.sh >/dev/null 2>&1\nfi\n/opt/slack_notify >/dev/null 2>&1\nif [ $GITHUB_DEPLOYMENT_STATUS == \"failure\" ];\nthen\n  exit 1\nelse\n  exit 0\nfi\n" in "/atlantis-data/repos/stone-payments/foundation-govcloud-infrastructure/799/prd/aws/applications/iam/policies/cross-account-policies": exit status 1: running "## check diferent paths to terraform in atlantis\nif [ -f \"/atlantis-data/bin/terraform$ATLANTIS_TERRAFORM_VERSION\" ]\nthen\n  /atlantis-data/bin/terraform$ATLANTIS_TERRAFORM_VERSION apply  $PLANFILE \nelse\n  /usr/local/bin/terraform$ATLANTIS_TERRAFORM_VERSION apply  $PLANFILE\nfi\n\nif [ $? -eq 0 ]; then\n  export GITHUB_DEPLOYMENT_STATUS=success\n  /opt/deployments\n  sh /opt/ddog-logs.sh >/dev/null 2>&1\nelse\n  export GITHUB_DEPLOYMENT_STATUS=failure\n  /opt/deployments\n  sh /opt/ddog-logs.sh >/dev/null 2>&1\nfi\n/opt/slack_notify >/dev/null 2>&1\nif [ $GITHUB_DEPLOYMENT_STATUS == \"failure\" ];\nthen\n  exit 1\nelse\n  exit 0\nfi\n" in "/atlantis-data/repos/stone-payments/foundation-govcloud-infrastructure/799/prd/aws/applications/iam/policies/cross-account-policies": 
module.cross-account-policy["ssmsessionmanagerrunshelly"].aws_cloudformation_stack_set_instance.stack_set_instance_accounts_application[0]: Creating...
╷
│ Error: Plugin did not respond
│ 
│   with module.cross-account-policy["ssmsessionmanagerrunshelly"].aws_cloudformation_stack_set_instance.stack_set_instance_accounts_application[0],
│   on ../../../../modules/stack-set/main.tf line 46, in resource "aws_cloudformation_stack_set_instance" "stack_set_instance_accounts_application":
│   46: resource "aws_cloudformation_stack_set_instance" "stack_set_instance_accounts_application" {
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵

Stack trace from the terraform-provider-aws_v5.62.0_x5 plugin:

panic: unexpected format for ID parts ([ssmsessionmanagerrunshellyb48c0a7221604959  us-east-1]), the following id parts indexes are blank ([1])

goroutine 362 [running]:
github.com/hashicorp/terraform-provider-aws/internal/errs.Must[...](...)
	github.com/hashicorp/terraform-provider-aws/internal/errs/must.go:13
github.com/hashicorp/terraform-provider-aws/internal/service/cloudformation.resourceStackSetInstanceCreate({0x174f43c8, 0xc0039964b0}, 0xc00116b780, {0x15222820, 0xc0026a8680})
	github.com/hashicorp/terraform-provider-aws/internal/service/cloudformation/stack_set_instance.go:272 +0x985
github.com/hashicorp/terraform-provider-aws/internal/provider.New.(*wrappedResource).Create.interceptedHandler[...].func8(0xc00116b780?, {0x15222820?, 0xc0026a8680})
	github.com/hashicorp/terraform-provider-aws/internal/provider/intercept.go:113 +0x283
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0x174f43c8?, {0x174f43c8?, 0xc0030afda0?}, 0xd?, {0x15222820?, 0xc0026a8680?})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.34.0/helper/schema/resource.go:801 +0x7a
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc000cae8c0, {0x174f43c8, 0xc0030afda0}, 0xc002fe4d00, 0xc00116b600, {0x15222820, 0xc0026a8680})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.34.0/helper/schema/resource.go:937 +0xa89
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc0036ad0e0, {0x174f43c8?, 0xc0030afce0?}, 0xc0006cb310)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.34.0/helper/schema/grpc_provider.go:1153 +0xd5c
github.com/hashicorp/terraform-plugin-mux/tf5muxserver.(*muxServer).ApplyResourceChange(0xc001967340, {0x174f43c8?, 0xc0030afa10?}, 0xc0006cb310)
	github.com/hashicorp/terraform-plugin-mux@v0.16.0/tf5muxserver/mux_server_ApplyResourceChange.go:36 +0x193
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc000f8c000, {0x174f43c8?, 0xc0030af050?}, 0xc001acc0e0)
	github.com/hashicorp/terraform-plugin-go@v0.23.0/tfprotov5/tf5server/server.go:865 +0x3d0
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x1500c680, 0xc000f8c000}, {0x174f43c8, 0xc0030af050}, 0xc00116ad80, 0x0)
	github.com/hashicorp/terraform-plugin-go@v0.23.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:518 +0x1a6
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0009f8200, {0x174f43c8, 0xc0030aefc0}, {0x17559500, 0xc000506000}, 0xc0030b0b40, 0xc00274eb70, 0x20d0b618, 0x0)
	google.golang.org/grpc@v1.63.2/server.go:1369 +0xdf8
google.golang.org/grpc.(*Server).handleStream(0xc0009f8200, {0x17559500, 0xc000506000}, 0xc0030b0b40)
	google.golang.org/grpc@v1.63.2/server.go:1780 +0xe8b
google.golang.org/grpc.(*Server).serveStreams.func2.1()
	google.golang.org/grpc@v1.63.2/server.go:1019 +0x8b
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 26
	google.golang.org/grpc@v1.63.2/server.go:1030 +0x125

Error: The terraform-provider-aws_v5.62.0_x5 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.

2024/08/13 09:27:03 INFO Deployment criado:  1718897256="missing value"
2024/08/13 09:27:03 INFO Deployment Status ID:  4310784725="missing value"

Dynamic environment variables added:
GOOGLE_CREDENTIALS

Dynamic environment variables added:
AWS_ACCESS_KEY_ID
AWS_SECRET_ACCESS_KEY
AWS_SESSION_TOKEN
```


### Terraform Configuration Files

I wrote an module, that we could use to deploy any stack_set template. It's simple and is composed by the following resources:

1. ``random_id``:  To create an random id to compose the stackset_name(cause it needs to be unique in the region)
2. ``aws_cloudformation_stack_set`` :  That create the stackset, based on an yml template, that has the content of an variable called, var.stackset.
3. ``aws_cloudformation_stack_set_instance "stack_set_instance_org_application" ``: Responsible for deployment of stackset in the target accounts in the entire organization
4. ``aws_cloudformation_stack_set_instance "stack_set_instance_accounts_application"``:  Responsible for deployment of stackset in the target accounts individualy.

I'm attaching the module created and the calling module that i've made to work with it.

[calling_module.zip](https://github.com/user-attachments/files/16599765/calling_module.zip)
[module-stack-set.zip](https://github.com/user-attachments/files/16599769/module-stack-set.zip)


### Steps to Reproduce

1. Get the module and the calling module and unpack it in some folder.
2. Configure the backend and provider credencials, in the calling module.
3. Change the reference for the source in the calling module, to get to the source of the module-stack-set
4. With all changes done, verify the .yml file in the ``cross-account-policies.d``, the var.stackset get the values from there
5. With everything ok, just plan in the folder of calling module.
6. The plan, must be ok
7. Then try to apply
8. You'll get the error that i've specified earlier.

### Debug Output

_No response_

### Panic Output

```console
running "## check diferent paths to terraform in atlantis\nif [ -f \"/atlantis-data/bin/terraform$ATLANTIS_TERRAFORM_VERSION\" ]\nthen\n  /atlantis-data/bin/terraform$ATLANTIS_TERRAFORM_VERSION apply  $PLANFILE \nelse\n  /usr/local/bin/terraform$ATLANTIS_TERRAFORM_VERSION apply  $PLANFILE\nfi\n\nif [ $? -eq 0 ]; then\n  export GITHUB_DEPLOYMENT_STATUS=success\n  /opt/deployments\n  sh /opt/ddog-logs.sh >/dev/null 2>&1\nelse\n  export GITHUB_DEPLOYMENT_STATUS=failure\n  /opt/deployments\n  sh /opt/ddog-logs.sh >/dev/null 2>&1\nfi\n/opt/slack_notify >/dev/null 2>&1\nif [ $GITHUB_DEPLOYMENT_STATUS == \"failure\" ];\nthen\n  exit 1\nelse\n  exit 0\nfi\n" in "/atlantis-data/repos/stone-payments/foundation-govcloud-infrastructure/799/prd/aws/applications/iam/policies/cross-account-policies": exit status 1: running "## check diferent paths to terraform in atlantis\nif [ -f \"/atlantis-data/bin/terraform$ATLANTIS_TERRAFORM_VERSION\" ]\nthen\n  /atlantis-data/bin/terraform$ATLANTIS_TERRAFORM_VERSION apply  $PLANFILE \nelse\n  /usr/local/bin/terraform$ATLANTIS_TERRAFORM_VERSION apply  $PLANFILE\nfi\n\nif [ $? -eq 0 ]; then\n  export GITHUB_DEPLOYMENT_STATUS=success\n  /opt/deployments\n  sh /opt/ddog-logs.sh >/dev/null 2>&1\nelse\n  export GITHUB_DEPLOYMENT_STATUS=failure\n  /opt/deployments\n  sh /opt/ddog-logs.sh >/dev/null 2>&1\nfi\n/opt/slack_notify >/dev/null 2>&1\nif [ $GITHUB_DEPLOYMENT_STATUS == \"failure\" ];\nthen\n  exit 1\nelse\n  exit 0\nfi\n" in "/atlantis-data/repos/stone-payments/foundation-govcloud-infrastructure/799/prd/aws/applications/iam/policies/cross-account-policies": 
module.cross-account-policy["ssmsessionmanagerrunshelly"].aws_cloudformation_stack_set_instance.stack_set_instance_accounts_application[0]: Creating...
╷
│ Error: Plugin did not respond
│ 
│   with module.cross-account-policy["ssmsessionmanagerrunshelly"].aws_cloudformation_stack_set_instance.stack_set_instance_accounts_application[0],
│   on ../../../../modules/stack-set/main.tf line 46, in resource "aws_cloudformation_stack_set_instance" "stack_set_instance_accounts_application":
│   46: resource "aws_cloudformation_stack_set_instance" "stack_set_instance_accounts_application" {
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵

Stack trace from the terraform-provider-aws_v5.62.0_x5 plugin:

panic: unexpected format for ID parts ([ssmsessionmanagerrunshellyb48c0a7221604959  us-east-1]), the following id parts indexes are blank ([1])

goroutine 362 [running]:
github.com/hashicorp/terraform-provider-aws/internal/errs.Must[...](...)
	github.com/hashicorp/terraform-provider-aws/internal/errs/must.go:13
github.com/hashicorp/terraform-provider-aws/internal/service/cloudformation.resourceStackSetInstanceCreate({0x174f43c8, 0xc0039964b0}, 0xc00116b780, {0x15222820, 0xc0026a8680})
	github.com/hashicorp/terraform-provider-aws/internal/service/cloudformation/stack_set_instance.go:272 +0x985
github.com/hashicorp/terraform-provider-aws/internal/provider.New.(*wrappedResource).Create.interceptedHandler[...].func8(0xc00116b780?, {0x15222820?, 0xc0026a8680})
	github.com/hashicorp/terraform-provider-aws/internal/provider/intercept.go:113 +0x283
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0x174f43c8?, {0x174f43c8?, 0xc0030afda0?}, 0xd?, {0x15222820?, 0xc0026a8680?})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.34.0/helper/schema/resource.go:801 +0x7a
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc000cae8c0, {0x174f43c8, 0xc0030afda0}, 0xc002fe4d00, 0xc00116b600, {0x15222820, 0xc0026a8680})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.34.0/helper/schema/resource.go:937 +0xa89
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc0036ad0e0, {0x174f43c8?, 0xc0030afce0?}, 0xc0006cb310)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.34.0/helper/schema/grpc_provider.go:1153 +0xd5c
github.com/hashicorp/terraform-plugin-mux/tf5muxserver.(*muxServer).ApplyResourceChange(0xc001967340, {0x174f43c8?, 0xc0030afa10?}, 0xc0006cb310)
	github.com/hashicorp/terraform-plugin-mux@v0.16.0/tf5muxserver/mux_server_ApplyResourceChange.go:36 +0x193
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc000f8c000, {0x174f43c8?, 0xc0030af050?}, 0xc001acc0e0)
	github.com/hashicorp/terraform-plugin-go@v0.23.0/tfprotov5/tf5server/server.go:865 +0x3d0
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x1500c680, 0xc000f8c000}, {0x174f43c8, 0xc0030af050}, 0xc00116ad80, 0x0)
	github.com/hashicorp/terraform-plugin-go@v0.23.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:518 +0x1a6
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0009f8200, {0x174f43c8, 0xc0030aefc0}, {0x17559500, 0xc000506000}, 0xc0030b0b40, 0xc00274eb70, 0x20d0b618, 0x0)
	google.golang.org/grpc@v1.63.2/server.go:1369 +0xdf8
google.golang.org/grpc.(*Server).handleStream(0xc0009f8200, {0x17559500, 0xc000506000}, 0xc0030b0b40)
	google.golang.org/grpc@v1.63.2/server.go:1780 +0xe8b
google.golang.org/grpc.(*Server).serveStreams.func2.1()
	google.golang.org/grpc@v1.63.2/server.go:1019 +0x8b
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 26
	google.golang.org/grpc@v1.63.2/server.go:1030 +0x125

Error: The terraform-provider-aws_v5.62.0_x5 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.

2024/08/13 09:27:03 INFO Deployment criado:  1718897256="missing value"
2024/08/13 09:27:03 INFO Deployment Status ID:  4310784725="missing value"
```

Dynamic environment variables added:
GOOGLE_CREDENTIALS

Dynamic environment variables added:
AWS_ACCESS_KEY_ID
AWS_SECRET_ACCESS_KEY
AWS_SESSION_TOKEN

### Important Factoids

I use atlantis tool to perform terraform plan and apply

### References

_No response_

### Would you like to implement a fix?

None

--- a/.changelog/38969.txt
+++ b/.changelog/38969.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+resource/aws_cloudformation_stack_set_instance: Fix crash during construction of the `id` attribute when `deployment_targets` does not include organizational unit IDs.
+```

--- a/internal/service/cloudformation/stack_set_instance.go
+++ b/internal/service/cloudformation/stack_set_instance.go
@@ -21,6 +21,7 @@ import (
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
 	"github.com/hashicorp/terraform-provider-aws/internal/conns"
+	"github.com/hashicorp/terraform-provider-aws/internal/create"
 	"github.com/hashicorp/terraform-provider-aws/internal/enum"
 	"github.com/hashicorp/terraform-provider-aws/internal/errs"
 	"github.com/hashicorp/terraform-provider-aws/internal/errs/sdkdiag"
@@ -32,6 +33,8 @@ import (
 )
 
 const (
+	ResNameStackSetInstance = "Stack Set Instance"
+
 	stackSetInstanceResourceIDPartCount = 3
 )
 
@@ -271,8 +274,12 @@ func resourceStackSetInstanceCreate(ctx context.Context, d *schema.ResourceData,
 		input.OperationPreferences = expandOperationPreferences(v.([]interface{})[0].(map[string]interface{}))
 	}
 
-	id := errs.Must(flex.FlattenResourceId([]string{stackSetName, accountOrOrgID, region}, stackSetInstanceResourceIDPartCount, false))
-	_, err := tfresource.RetryWhen(ctx, propagationTimeout,
+	id, err := flex.FlattenResourceId([]string{stackSetName, accountOrOrgID, region}, stackSetInstanceResourceIDPartCount, false)
+	if err != nil {
+		return create.AppendDiagError(diags, names.CloudFormation, create.ErrActionFlatteningResourceId, ResNameStackSetInstance, id, err)
+	}
+
+	_, err = tfresource.RetryWhen(ctx, propagationTimeout,
 		func() (interface{}, error) {
 			input.OperationId = aws.String(sdkid.UniqueId())
 

--- a/internal/service/cloudformation/stack_set_instance.go
+++ b/internal/service/cloudformation/stack_set_instance.go
@@ -249,7 +249,9 @@ func resourceStackSetInstanceCreate(ctx context.Context, d *schema.ResourceData,
 
 	if v, ok := d.GetOk("deployment_targets"); ok && len(v.([]interface{})) > 0 && v.([]interface{})[0] != nil {
 		dt := expandDeploymentTargets(v.([]interface{}))
-		accountOrOrgID = strings.Join(dt.OrganizationalUnitIds, "/")
+		if len(dt.OrganizationalUnitIds) > 0 {
+			accountOrOrgID = strings.Join(dt.OrganizationalUnitIds, "/")
+		}
 		input.DeploymentTargets = dt
 	} else {
 		d.Set(names.AttrAccountID, accountID)

